python/ray/train/v2/_internal/execution/controller/controller.py

```diff
@@ -216,39 +219,33 @@ def _execute_failure_decision(
             return TrainControllerLoopIterationResult(
                 run_attempt_id=self._get_run_attempt_id(),
                 previous_state=controller_state,
-                next_state=RunningState(),
+                next_state=controller_state,
+                error=error,
             )
 
-        errors_str = worker_group_status.get_error_string()
-        training_failed_error = TrainingFailedError(
-            error_message=errors_str, worker_failures=worker_group_status.errors
-        )
-
         if failure_decision == FailureDecision.RESTART:
-            logger.error(
-                "Restarting training worker group after encountering "
-                f"failures on {len(worker_group_status.errors)} worker(s):\n"
-                f"{errors_str}"
+            assert isinstance(error, TrainingFailedError)
+            return TrainControllerLoopIterationResult(
+                run_attempt_id=self._get_run_attempt_id(),
+                previous_state=controller_state,
+                next_state=RestartingState(training_failed_error=error),
             )
-            next_state = RestartingState(training_failed_error=training_failed_error)
+        elif failure_decision == FailureDecision.RESCHEDULE:
+            assert isinstance(error, ControllerError)
             return TrainControllerLoopIterationResult(
                 run_attempt_id=self._get_run_attempt_id(),
                 previous_state=controller_state,
-                next_state=next_state,
-                training_failed_error=training_failed_error,
+                next_state=ReschedulingState(controller_failed_error=error),
             )
```

Comment:
In the current implementation there is a 1:1 mapping between `TrainingFailedError`:`RESTART` and `ControllerError`:`RESCHEDULE`, but I don't think we want to enforce this - generally the `FailurePolicy` should own the entirety of the decision making/validation logic. I can imagine that in the future there are cases where `ControllerError` can return `RESTART` as well.